File: core/src/com/tiles/pathfinding/Tile.java
package com.tiles.pathfinding;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Input;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.graphics.g2d.BitmapFont;
import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.utils.Json;

import java.io.Serializable;

import static com.tiles.pathfinding.NeededConstants.*;
import static jdk.internal.dynalink.support.Guards.isNull;

public class Tile implements Serializable {
    public int number; // numéro de tuile
    private String path; // path de la tuile
    private transient Sprite sprite; // transient ça veut dire qu'on le stock pas dans la serialization

    public Sprite getSprite() {
        return sprite;
    } // On en a besoin pour la pile

    public Case[][] caseList; // Un tableau de 4x4 avec les cases
    private Case[][][] caseListofCases; // Pour la serialization
    public int rotation = 0; // Indicateur de rotation (dans le sens trigonométrique)
    public boolean[] exits;
    public Case[] exitCases;
    // Exit représente où sont les sorties de cette tuile
    // Comme d'habitude, 0 est le Sud, et on tourne dans le sens trigo
    public int entrance;
    public Case entranceCase;
    // L'entrée représente la direction cardinale de la tuile

    private long cooldown;

    public void startCooldown() {
        cooldown = System.currentTimeMillis();
    }
    // L'utilité de cette variable et de cette méthode est questionnable
    // Elles sont utiles pour éviter un phénomène que j'appelle le blinking
    // Qui fait que, au moment où on dépose la carte, les trucs du pathfinding apparaissent


    public float x = 0;
    public float y = 0;
    private float size = NeededConstants.tileSize;

    public float getX() {
        return x;
    }

    public float getY() {
        return y;
    }

    public void updateAll() {
        sprite.setX(x);
        sprite.setY(y);
        updateSize();
    }

    private void updateCoordinates() {
        sprite.setX(x);
        sprite.setY(y);
        updateAllCases();
    }

    public float getSize() {
        return size;
    }

    public void updateSize() {
        sprite.setSize(size, size);
        updateAllCases();
    }

    // fonctions classiques j'ai envie de dire
    private boolean exploring = false;
    public void handleInput(Player player, BitmapFont numberCase) {
        Case tempCase;
        Vector2 mouseInput = mouseInput();
        // Puis si on clique gauche, boum, le pathfinding
        if (!exploring && (Gdx.input.isButtonJustPressed(Input.Buttons.LEFT)) && (System.currentTimeMillis() - cooldown > 500)) {
            try {
                tempCase = getCase(mouseInput);
                tempCase.show();
                tempCase.explore(player);
                numberCase.draw(NeededConstants.batch, "x = " + tempCase.x + "; y = " + tempCase.y + "; couleur = " + tempCase.color + ", portal = " + tempCase.hasPortal, 700f, 200f);
                exploring = true;
                lastExploredCase = tempCase;
            } catch (ArrayIndexOutOfBoundsException e) {
            }
            // Bah oui parce que si on est pas dans les bornes de la tuile forcément getCase fonctionne moins bien lol
        }
        else if (exploring && (Gdx.input.isButtonJustPressed(Input.Buttons.LEFT))) {
            lastExploredCase.revert(player);
            lastExploredCase.hide();
            exploring = false;
        }
        // On gère la rotation
    }


    Tile(int number) {
        this.number = number;
        // Bon on stock le path quelque part quand même
        path = "tuiles/tile" + number + ".jpg";
        /* Et maintenant, on construit allégrement la tuile
        Je vais maintenant expliquer comment on utilise le système de tableau
        On a un tableau 4x4 pour les cases, et pour les murs...
        Un tableau de 3x4 pour ceux verticaux
        Un tableau de 4x3 pour ceux horizontaux
        Pour le mur: 1 = un mur, 0 = pas de mur
        Pour les cases:
         - 0 = innacessible
         - 1 = accessible, case normale
         - 2 = une entrée
         // les racourcis et les escalators sont rajoutés à la main
         - 7 = un sablier
         - 8 = boule de crystal
         - 9 = caméra
         - 1x = une sortie normale
         - 2x = un téléporteur (voir comment les coder plus tard)
         - 3x = une arme
         - 4x = une sortie de joueur

         pour les grands nombres, on a
         - 0 -> vert
         - 1 -> violet
         - 2 -> jaune
         - 3 -> orange
         */
        caseList = new Case[4][4];
        caseListofCases = new Case[4][4][4];

        // On commence par créer le tableau de cases avec des Case à l'intérieur
        for (int tempX = 0; tempX < 4; tempX++) {
            for (int tempY = 0; tempY < 4; tempY++) {
                caseList[tempY][tempX] = new Case(TileArray.getArray(number)[tempY][tempX], this);
                caseListofCases[tempY][tempX] = caseList[tempY][tempX].caseList;
            }
        }
        // Puis on créé les liaisons entre les cases
        for (Case[] ligne : caseList) {
            for (Case tempCase : ligne) {
                tempCase.getNeighbours(
                        TileArray.getArrayWallHorizontal(number),
                        TileArray.getArrayWallVertical(number));
            }
        }

        exitCases = new Case[]{caseList[0][1], caseList[1][3],caseList[3][2], caseList[2][0]};
        exits = new boolean[]{caseList[0][1].isExit, caseList[1][3].isExit,caseList[3][2].isExit, caseList[2][0].isExit};

        System.out.println("Tile number " + number + " has exits " + exits[0] + " "+ exits[1] + " "+ exits[2] + " "+ exits[3]);

        if (caseList[0][1].isEntrance) {
            entrance = 0;
            entranceCase = caseList[0][1];
        }
        if (caseList[1][3].isEntrance) {
            entrance = 1;
            entranceCase = caseList[1][3];
        }
        if (caseList[2][0].isEntrance) {
            entrance = 3;
            entranceCase = caseList[2][0];
        }
        if (caseList[3][2].isEntrance) {
            entrance = 2;
            entranceCase = caseList[3][2];
        }
        System.out.println("Tile number " + number + " has entrance number " + entrance);
        // On récupère la coordonée d'entrée
    }

    public void load() { // Obligatoire pour la serialization
        sprite = new Sprite(new Texture(path)); // On se charge soit même
        sprite.setOrigin(tileSize/2, tileSize/2);
        for (int i = 0; i <=3; i ++) {
            for (int j = 0; j <=3; j++)
                caseList[j][i].load(this, caseListofCases[j][i]); // et on charge toutes les cases
        }
        // Et on rajoute les raccourcis et escalators
        complete();
        updateAll();
    }

    public void draw() {
        sprite.draw(NeededConstants.batch);
        for (Case[] ligne : caseList) {
            for (Case tempCase : ligne)
                tempCase.draw();
        }
    }

    public void showAll(Batch batch) { // Truc de déboguage
        for (Case[] ligne : caseList) {
            for (Case tempCase : ligne)
                tempCase.show();
        }
    }

    public int[] getCaseCoordinates(Case self) {
        // C'est juste pour récupérer le x,y d'une case parce que flemme de le faire autrement
        int[] xy = {0, 0};
        for (Case[] ligne : caseList) {
            xy[0] = 0;
            for (Case comparator : ligne) {
                if (comparator.equals(self)) return xy;
                xy[0]++;
            }
            xy[1]++;
        }
        return xy;
    }

    public Case getCase(Vector2 mouseInput) {
        int tempX = (int) Math.floor((mouseInput.x - x - offset) / caseSize);
        int tempY = (int) Math.floor((mouseInput.y - y - offset ) / caseSize);
        int buffer;
        if (rotation == 1) {
            buffer = tempX;
            tempX = tempY;
            tempY = 3 - buffer;
        }
        if (rotation == 2) {
            tempX = 3 - tempX;
            tempY = 3 - tempY;
        }
        if (rotation == 3) {
            buffer = tempX;
            tempX = 3 - tempY;
            tempY = buffer;
        }
        return caseList[tempY][tempX];
        // C'est du calcul simple, si tu comprends pas retourne en maternelle
    }


    public void rotate(int angle) {
        rotation += angle;
        rotation = (rotation % 4 + 4) % 4; // Java et les modulos...
        sprite.rotate(angle * 90); // Dans le sens trigo
        updateAllCases();
    }

    private void updateAllCases() {
        for (Case[] ligne : caseList) {
            for (Case tempCase : ligne) {
                tempCase.updateCoordinates();
                tempCase.setSize(128 * sprite.getWidth() / 600);
            }
        }
    }


    private void complete() {
        // Petite fonction pour rajouter les escalators et les raccourcis
        if (number == 2) Case.makeElevator(caseList[2][0], caseList[0][1]);
        if (number == 7) Case.makeElevator(caseList[0][1], caseList[2][2]);
        if (number == 10) Case.makeElevator(caseList[2][2], caseList[1][1]);
        if (number == 12) {
            Case.makeElevator(caseList[3][1], caseList[2][2]);
            Case.makeElevator(caseList[1][0], caseList[0][1]);
        }
        if (number == 13) Case.makeShortcut(caseList[2][2], caseList[2][3]);
        if (number == 14) {
            Case.makeElevator(caseList[2][0], caseList[1][2]);
            Case.makeShortcut(caseList[1][2], caseList[0][2]);
        }
        if (number == 15) {
            Case.makeElevator(caseList[3][0], caseList[2][2]);
            Case.makeShortcut(caseList[1][0], caseList[1][1]);
        }
        if (number == 16) Case.makeShortcut(caseList[1][2], caseList[1][3]);
        if (number == 19) Case.makeShortcut(caseList[1][2], caseList[1][3]);
        if (number == 24) Case.makeShortcut(caseList[0][2], caseList[0][3]);
    }

    public void dispose() { // fonction dispose classique
        sprite.getTexture().dispose();
        for (Case[] ligne : caseList) {
            for (Case tempCase : ligne)
                tempCase.dispose();
        }
    }

    private static boolean isValidPlacement(Tile tileToPlace, Tile tileToJoin) {
        if (tileToJoin == null) return false;
        return tileToJoin.exits[((2+tileToPlace.entrance - tileToPlace.rotation + tileToJoin.rotation)%4 +4)%4];
    }

    private Tile[] getNeighbouringTiles() {
        Vector2 mousePosition = mouseInput();
        mousePosition.sub(NeededConstants.origin);
        mousePosition.sub(tileSize/2, tileSize/2);
        mousePosition.mul(NeededConstants.newBaseInvert);
        int x = Math.round(mousePosition.x);
        int y = Math.round(mousePosition.y);
        return new Tile[]{
                getTile(new Vector2(x,y-1).mul(NeededConstants.newBase).add(NeededConstants.origin).add(tileSize/2, tileSize/2)),
                getTile(new Vector2(x+1,y).mul(NeededConstants.newBase).add(NeededConstants.origin).add(tileSize/2, tileSize/2)),
                getTile(new Vector2(x,y+1).mul(NeededConstants.newBase).add(NeededConstants.origin).add(tileSize/2, tileSize/2)),
                getTile(new Vector2(x-1,y).mul(NeededConstants.newBase).add(NeededConstants.origin).add(tileSize/2, tileSize/2)),
        };
    }

    private boolean noOverlap() {
        for (Tile tile : tileList) {
            if (!tile.equals(this) && ((tile.x < mouseInput().x) && (mouseInput().x < tile.x + tileSize) && (tile.y < mouseInput().y) && (mouseInput().y < tile.y + tileSize))) {
                return false;
            }
        }
        return true;
    }
    public boolean canPlaceThere() {
        Tile[] neighbors = getNeighbouringTiles();
        int index = ((entrance+rotation)%4 + 4)%4;
        boolean temp = isValidPlacement(this, neighbors[index]) && noOverlap();
        if (temp) {
            entranceCase.caseList[index] = neighbors[index].exitCases[(index+2) % 4];
            neighbors[index].exitCases[(index+2)%4].caseList[(index+2)%4] = entranceCase;
            Tile tempTile;
            int indexPrime;
//            for (int i = 0; i <= 3; i ++) {
//                tempTile = neighbors[i];
//                index = ((i+rotation)%4 + 4)%4;
//                if (exits[index] &&  tempTile != null && tempTile.exits[((2+i - tempTile.rotation)%4 +4)%4]) {
//                    indexPrime = ((2+i - tempTile.rotation)%4 +4)%4;
//                    exitCases[index].caseList[i] = tempTile.exitCases[indexPrime];
//                    tempTile.exitCases[indexPrime].caseList[indexPrime] = exitCases[index];
//                }
//            }
        }
        return temp;
    }
}

File: core/src/tsp/genint/screens/game/board/Tile.kt
package tsp.genint.screens.game.board

import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.math.Vector2
import tsp.genint.screens.GameScreens
import tsp.genint.screens.game.BaseActor
import tsp.genint.screens.game.board.Case.Companion.link
import tsp.genint.utils.*
import java.io.Serializable
import kotlin.math.floor
import kotlin.math.roundToInt

class Tile internal constructor(val number: Int) : Serializable {
    private val path = "Game/Tiles/tile$number.png"

    @Transient
    var sprite: BaseActor? = null
        private set
    var caseList: List<List<Case>>
    private val caseListofCases: List<List<List<Case?>>>
    var rotation = 0 // Indicateur de rotation (dans le sens trigonométrique)
    var exits: BooleanArray
    var exitCases: List<Case>
    private var entrance = Directions.SOUTH
    private var entranceCase: Case? = null

    private var cooldown: Long = 0
    fun startCooldown() {
        cooldown = System.currentTimeMillis()
    }

    var x = 0f
    var y = 0f
    val size = tileSize
    fun updateAll() {
        sprite!!.x = x
        sprite!!.y = y
        updateSize()
    }

    private fun updateSize() {
        sprite!!.setSize(size, size)
        updateAllCases()
    }

    init {
        // On commence par créer le tableau de cases avec des Case à l'intérieur
        val tileArray = getTileArray(number)
        caseList = List(4) { tempY -> List(4) { tempX -> Case(tileArray[tempY][tempX], this) } }
        caseListofCases = List(4) { tempY -> List(4) { tempX -> caseList[tempY][tempX].caseList } }

        // Puis on créé les liaisons entre les cases
        val horizontal = getTileArrayWallHorizontal(number)
        val vertical = getTileArrayWallVertical(number)
        for (ligne in caseList) {
            for (tempCase in ligne) {
                tempCase.getNeighbours(horizontal, vertical)
            }
        }
        exitCases = listOf(caseList[0][1], caseList[1][3], caseList[3][2], caseList[2][0])
        exits = booleanArrayOf(caseList[0][1].exit, caseList[1][3].exit, caseList[3][2].exit, caseList[2][0].exit)

        if (caseList[0][1].entrance) {
            entrance = Directions.SOUTH
            entranceCase = caseList[0][1]
        }
        if (caseList[1][3].entrance) {
            entrance = Directions.EAST
            entranceCase = caseList[1][3]
        }
        if (caseList[2][0].entrance) {
            entrance = Directions.WEST
            entranceCase = caseList[2][0]
        }
        if (caseList[3][2].entrance) {
            entrance = Directions.NORTH
            entranceCase = caseList[3][2]
        }
        complete()
    }

    fun load() { // Obligatoire pour la serialization
        sprite = BaseActor(Texture(path)) // On se charge soit même
        sprite!!.setOrigin(tileSize / 2, tileSize / 2)
        GameScreens.gameScreen.mainStage.addActor(sprite)
        for (i in 0..3) {
            for (j in 0..3) caseList[j][i].load(this, caseListofCases[j][i]) // et on charge toutes les cases
        }
        // Et on rajoute les raccourcis et escalators
        updateAll()
    }

    fun getCaseCoordinates(self: Case): IntArray {
        // C'est juste pour récupérer le x,y d'une case parce que flemme de le faire autrement
        // Sous entendu, x et y valent 0,1,2 ou 3, c'est leur position dans caseList
        val xy = intArrayOf(0, 0)
        for (ligne in caseList) {
            xy[0] = 0
            for (comparator in ligne) {
                if (comparator == self) return xy
                xy[0]++
            }
            xy[1]++
        }
        return xy
    }

    fun getCase(mouseInput: Vector2): Case {
        // Pour récupérer la case située à une certaine coordonée
        var tempX = floor(((mouseInput.x - x - offset) / caseSize).toDouble()).toInt()
        var tempY = floor(((mouseInput.y - y - offset) / caseSize).toDouble()).toInt()
        var buffer: Int
        if (rotation == 1) {
            buffer = tempX
            tempX = tempY
            tempY = 3 - buffer
        }
        if (rotation == 2) {
            tempX = 3 - tempX
            tempY = 3 - tempY
        }
        if (rotation == 3) {
            buffer = tempX
            tempX = 3 - tempY
            tempY = buffer
        }
        return caseList[tempY][tempX]
    }

    fun rotate(angle: Int) {
        rotation += angle
        rotation = modulo(rotation, 4)
        sprite!!.rotateBy((angle * 90).toFloat()) // Dans le sens trigo
        updateAllCases()
    }

    private fun updateAllCases() {
        for (ligne in caseList) {
            for (tempCase in ligne) {
                tempCase.updateCoordinates()
                tempCase.setSize(128 * sprite!!.width / 600)
            }
        }
    }

    private fun complete() {
        // Petite fonction pour rajouter les escalators et les raccourcis
        if (number == 2) Case.makeElevator(caseList[2][0], caseList[0][1])
        if (number == 7) Case.makeElevator(caseList[0][1], caseList[2][2])
        if (number == 10) Case.makeElevator(caseList[2][2], caseList[1][1])
        if (number == 12) {
            Case.makeElevator(caseList[3][1], caseList[2][2])
            Case.makeElevator(caseList[1][0], caseList[0][1])
        }
        if (number == 13) Case.makeShortcut(caseList[2][2], caseList[2][3])
        if (number == 14) {
            Case.makeElevator(caseList[2][0], caseList[1][2])
            Case.makeShortcut(caseList[1][2], caseList[0][2])
        }
        if (number == 15) {
            Case.makeElevator(caseList[3][0], caseList[2][2])
            Case.makeShortcut(caseList[1][0], caseList[1][1])
        }
        if (number == 16) Case.makeShortcut(caseList[1][2], caseList[1][3])
        if (number == 19) Case.makeShortcut(caseList[1][2], caseList[1][3])
        if (number == 24) Case.makeShortcut(caseList[0][2], caseList[0][3])

        // C'est du bon gros nesting mais bon, ça fonctionne, et c'est """compact"""
    }

    fun dispose() {
        sprite!!.remove()
        for (ligne in caseList) {
            for (tempCase in ligne) tempCase.dispose()
        }
    }

    private val neighbouringTiles: Array<Tile?>
        get() {
            val mousePosition = Vector2(sprite!!.x, sprite!!.y)
            mousePosition.sub(origin)
            //        mousePosition.sub(getTileSize( / 2, getTileSize( / 2);
            mousePosition.mul(newBaseInvert)
            val x = mousePosition.x.roundToInt()
            val y = mousePosition.y.roundToInt()
            return arrayOf( //South
                getTile(
                    Vector2(x.toFloat(), (y - 1).toFloat()).mul(newBase).add(origin).add(tileSize / 2, tileSize / 2)
                ),  // East
                getTile(
                    Vector2((x + 1).toFloat(), y.toFloat()).mul(newBase).add(origin).add(tileSize / 2, tileSize / 2)
                ),  // North
                getTile(
                    Vector2(x.toFloat(), (y + 1).toFloat()).mul(newBase).add(origin).add(tileSize / 2, tileSize / 2)
                ),  // West
                getTile(
                    Vector2((x - 1).toFloat(), y.toFloat()).mul(newBase).add(origin).add(tileSize / 2, tileSize / 2)
                )
            )
        }

    private fun noOverlap(): Boolean {
        for (tile in tileList) {
            if (tile != this && tile.x < mouseInput().x && mouseInput().x < tile.x + tileSize && tile.y < mouseInput().y && mouseInput().y < tile.y + tileSize) {
                return false
            }
        }
        return true
    }

    fun canPlaceThere(): Boolean {
        val neighbors = neighbouringTiles
        val direction = modulo(entrance + rotation, Directions.values().size)
        return isValidPlacement(neighbors[direction], direction) && noOverlap()
    }

    fun place() {
        val neighbors = neighbouringTiles
        var direction = modulo(entrance + rotation, Directions.values().size)
        if (isValidPlacement(neighbors[direction], direction) && noOverlap()) {
            link(
                entranceCase!!,
                neighbors[direction]!!.exitCases[modulo(
                    direction + 2 - neighbors[direction]!!.rotation,
                    Directions.values().size
                )],
                direction
            )
            // Puis prise en charge des exits et du link des exits
            for (i in Directions.values().indices) {
                if (exits[i]) { // On commence par vérifier que i est bien une sortie
                    direction = modulo(i + rotation, Directions.values().size)
                    if (neighbors[direction] != null && neighbors[direction]!!.exits[modulo(
                            direction + 2 - neighbors[direction]!!.rotation,
                            Directions.values().size
                        )]
                    ) {
                        // On fait la même chose qu'avant, mais on considère simplement
                        // notre exit comme une entrance
                        link(
                            exitCases[i],
                            neighbors[direction]!!.exitCases[modulo(
                                direction + 2 - neighbors[direction]!!.rotation,
                                Directions.values().size
                            )],
                            direction
                        )
                    }
                }
            }
        }
        for (ligne in caseList) {
            for (tempCase in ligne) {
                if (tempCase.hasPortal) {
                    portalList[tempCase.color]!!.add(tempCase)
                }
            }
        }
    }

    companion object {
        private fun isValidPlacement(tileToJoin: Tile?, direction: Int): Boolean {
            // direction corresponds à la direction de tileToJoin
            // Relativement à la tile que l'on cherche à poser
            // On vérifie simplement qu'il y a bien une case, qu'il s'agit bien d'une sortie
            // Et, le cas échéant, qu'il y a bien un pion de la bonne couleur
            if (tileToJoin != null && tileToJoin.exits[modulo(
                    direction + 2 - tileToJoin.rotation,
                    Directions.values().size
                )]
            ) {
                val exitCase =
                    tileToJoin.exitCases[modulo(direction + 2 - tileToJoin.rotation, Directions.values().size)]
                return exitCase.pawn != null && exitCase.color === exitCase.pawn!!.color
            }
            return false
        }
    }
}
